qbraid_qir/squin/visitor.py

```diff
@@ -0,0 +1,562 @@
+                                match = re.search(
+                                    r"inttoptr \(i64 (\d+) to %Result\*\)", operand_str
+                                )
+                                if match:
+                                    cbit_indices.append(int(match.group(1)))
+                                elif "null" in operand_str:
+                                    cbit_indices.append(0)
+
+                            # Extract parameters (double values)
+                            elif "double" in operand_str:
+                                match = re.search(r"double ([0-9.e+-]+)", operand_str)
+                                if match:
+                                    parameters.append(float(match.group(1)))
+                                else:
+                                    # Handle hex double values
+                                    hex_match = re.search(r"double 0x([0-9a-fA-F]+)", operand_str)
+                                    if hex_match:
+                                        try:
+                                            int_val = int(hex_match.group(1), 16)
+                                            float_val = struct.unpack(
+                                                "d", struct.pack("Q", int_val)
+                                            )[0]
+                                            parameters.append(float_val)
+                                        except:
+                                            parameters.append(f"0x{hex_match.group(1)}")
+
+                        gate_info = {
+                            "gate_name": gate_name,
+                            "qubit_indices": qubit_indices,
+                            "cbit_indices": cbit_indices,
+                            "parameters": parameters,
+                        }
+
+                        gates.append(gate_info)
+
+        return gates
+
+    def get_num_qubits(self, module):
+        """
+        Extract the number of qubits from the pyqir module.
+
+        Args:
+            module: PyQIR module object
+
+        Returns:
+            int: Number of qubits
+        """
+        # Find the entry point function
+        entry_func = None
+        for func in module.functions:
+            if pyqir.is_entry_point(func):
+                entry_func = func
+                break
+
+        if not entry_func:
+            return 0
+
+        # Get the required number of qubits and results (classical bits)
+        qubits = pyqir.required_num_qubits(entry_func)
+        num_cbits = pyqir.required_num_results(entry_func)
+
+        return qubits
+
+    def get_gate_operation(self, instruction: pyqir.Instruction):
+        str_instruction = str(instruction)
+        if "call void @__quantum__qis__" in str_instruction and "__body(" in str_instruction:
+            gate_match = re.search(r"@__quantum__qis__(\w+)__body", str_instruction)
```

Comment:
Comment for this and other string processing operation in the code below - is there any kind of object structure through which we can extract this info directly? Regular expression search can get very expensive pretty quickly